runtime-transaction/src/compute_budget_program_id_filter.rs

```diff
@@ -1,37 +1,152 @@
+}
+
 pub(crate) struct ComputeBudgetProgramIdFilter {
     // array of slots for all possible static and sanitized program_id_index,
     // each slot indicates if a program_id_index has not been checked (eg, None),
-    // or already checked with result (eg, Some(result)) that can be reused.
-    flags: [Option<bool>; FILTER_SIZE as usize],
+    // or already checked with result (eg, Some(ProgramKind)) that can be reused.
+    program_kind: [Option<ProgramKind>; FILTER_SIZE as usize],
 }
 
 impl ComputeBudgetProgramIdFilter {
     pub(crate) fn new() -> Self {
         ComputeBudgetProgramIdFilter {
-            flags: [None; FILTER_SIZE as usize],
+            program_kind: [None; FILTER_SIZE as usize],
         }
     }
 
     #[inline]
     pub(crate) fn is_compute_budget_program(&mut self, index: usize, program_id: &Pubkey) -> bool {
+        // Access the program kind at the given index, panic if index is invalid.
+        match self
+            .program_kind
+            .get(index)
+            .expect("program id index is sanitized")
+        {
+            // If the program kind is already set, check if it matches the target ProgramKind.
+            Some(program_kind) => {
+                *program_kind
+                    == ProgramKind::Builtin {
+                        is_compute_budget: true,
+                    }
+            }
+            // If the program kind is not set, calculate it, store it, and then check the condition.
+            None => {
+                let is_compute_budget = Self::check_compute_bugdet_program_id(program_id);
+
+                if is_compute_budget {
+                    self.program_kind[index] = Some(ProgramKind::Builtin {
+                        is_compute_budget: true,
+                    });
+                }
+
+                is_compute_budget
+            }
+        }
+    }
+
+    #[inline]
+    pub(crate) fn get_program_kind(&mut self, index: usize, program_id: &Pubkey) -> ProgramKind {
         *self
-            .flags
+            .program_kind
             .get_mut(index)
             .expect("program id index is sanitized")
-            .get_or_insert_with(|| Self::check_program_id(program_id))
+            .get_or_insert_with(|| Self::check_program_kind(program_id))
     }
 
     #[inline]
-    fn check_program_id(program_id: &Pubkey) -> bool {
+    fn check_compute_bugdet_program_id(program_id: &Pubkey) -> bool {
         if !MAYBE_BUILTIN_KEY[program_id.as_ref()[0] as usize] {
             return false;
         }
+
         solana_sdk::compute_budget::check_id(program_id)
     }
+
+    #[inline]
+    fn check_program_kind(program_id: &Pubkey) -> ProgramKind {
+        if !MAYBE_BUILTIN_KEY[program_id.as_ref()[0] as usize] {
+            return ProgramKind::NotBuiltin;
+        }
+
+        if is_builtin_program(program_id) {
+            ProgramKind::Builtin {
+                is_compute_budget: solana_sdk::compute_budget::check_id(program_id),
```

Comment:
nit: technically we would know that this isn't the compute budget program already because otherwise the filter would have had a compute budget builtin entry that was populated by the first filter pass